README.md
## Sha alerts - Alert dialogs for flutter with UX in mind.

[Sha alert](https://pub.dev/packages/sha_alert) is alert dialogs designed by considering user experience and user interface in first place.
<br>
<br>


<div align='center'>

<img width='30%' src='https://i.ibb.co/FmjzMSm/hero-modified.png'>
</div>


<br>


Sha alert supports 5 types of alert dialogs.
<b> - Primary

- Secondary
- Positive
- Negative
- Warning

</b>

### Features and UX

- Haptic feedbacks on supported devices.
- Naturally animated (Not starting scale from 0 to 1 instead 0.7 to 1).
- Scale animation + alpha animation.
- Two action buttons.
- Customizable title, message, action buttons and colors.

<b> PRO TIP 🚀 : Read my article on fueler on how to design alert dialogs [here](https://fueler.io/shajin/how-to-design-alert-dialogs-ux-flutter)
<b>

## Screenshots

<div style="display: flex; flex-wrap: wrap;">
    <img width='30%' src="https://i.ibb.co/Mfn3NWv/Whats-App-Image-2023-07-30-at-12-34-05-PM-2.jpg"/>
    <img width='30%' src="https://i.ibb.co/7YN7j6x/Simulator-Screenshot-i-Phone-14-Pro-2023-07-30-at-17-49-42.png"/>
    <img width='30%' src="https://i.ibb.co/tB1Y2fK/Simulator-Screenshot-i-Phone-14-Pro-2023-07-30-at-18-00-26.png"/>
    <img width='30%' src="https://i.ibb.co/c6g10QX/Simulator-Screenshot-i-Phone-14-Pro-2023-07-30-at-17-57-25.png"/>
</div>

## Installation

```
dependencies:
  sha_alert: ^0.0.1
```

## Usage

```dart

ShaAlert.showHozzoAlert(
  context: context,
  title: 'Welcome Note',
  message: 'Happy to see you on board! Please read the terms and conditions carefully.',
  firstButtonTitle: "I Agree",
  firstButtonType: ButtonType.primary,
  onFirstButtonPressed: () {
    // Code to be executed when the first button is pressed
  },
  secondButtonTitle: "Read Terms and Conditions",
  onSecondButtonPressed: () {
    // Code to be executed when the second button is pressed
  },
);

```



### Parameters:

1. `context` (BuildContext) - *(Required)* This context should contain the navigator for the alert to work properly and pop itself.

2. `title` (String) - *(Required)* The title of the alert dialog.

3. `message` (String) - *(Required)* The main message/content of the alert dialog.

4. `firstButtonTitle` (String) - *(Required)* The title of the first action button.

5. `secondButtonTitle` (String, optional) - The title of the second action button. If not provided, only one button will be shown.

6. `onFirstButtonPressed` (Function, optional) - Function to be called when the first action button is pressed.

7. `onSecondButtonPressed` (Function, optional) - Function to be called when the second action button is pressed.

8. `firstButtonType` (ButtonType, optional) - The type of the first action button. Default is `ButtonType.negative`. Possible values are `ButtonType.negative`, `ButtonType.positive`, and `ButtonType.secondary`.

9. `secondButtonType` (ButtonType, optional) - The type of the second action button. Default is `ButtonType.secondary`. Possible values are `ButtonType.negative`, `ButtonType.positive`, and `ButtonType.secondary`.

10. `height` (double, optional) - Height of the alert dialog. Default is 280 pixels.

11. `child` (Widget, optional) - Use this for showing custom widgets or even images. This can be used to add additional content to the alert dialog. Make sure to set the `height` accordingly if you use this parameter.

12. `backdropColor` (Color, optional) - Backdrop color of the alert dialog. Default is `Colors.black54` (black with 54% opacity).

13. `transitionDuration` (Duration, optional) - Default transition duration for the alert dialog. Default is 120 milliseconds (0.12 seconds).

14. `onDismissed` (Function, optional) - Function to be called when the alert is dismissed (e.g., when the user taps outside the dialog).

15. `customHapticFeedback` (Function, optional) - Custom haptic feedback function to be called when the alert is displayed. If not provided, default haptic feedback will be used. This function will be called after 100 milliseconds.

16. `secondaryColor` (Color, optional) - Custom color for secondary elements in the alert dialog (e.g., button backgrounds). If `overrideTheme` is set to `false`, this color will be set to the secondary color of the current theme. Otherwise, the default is `Colors.grey.shade100`.

17. `primaryColor` (Color, optional) - Custom color for primary elements in the alert dialog (e.g., button text). If `overrideTheme` is set to `false`, this color will be set to the primary color of the current theme. Otherwise, the default is `Colors.blue`.

18. `errorColor` (Color, optional) - Custom color for error elements in the alert dialog (e.g., error button backgrounds). If `overrideTheme` is set to `false`, this color will be set to the error color of the current theme. Otherwise, the default is `Colors.red`.

19. `successColor` (Color, optional) - Custom color for success elements in the alert dialog (e.g., success button backgrounds). If `overrideTheme` is set to `false`, this color will be set to the primary color of the current theme. Otherwise, the default is `Colors.green`.

20. `overrideTheme` (bool, optional) - If set to `true`, the custom colors specified (`primaryColor`, `secondaryColor`, `errorColor`, `successColor`) will be used. If set to `false`, the colors will be set based on the current theme.




### Author
Shajin KP - [shajin.in](https://shajin.in)


### Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.


CHANGELOG.md
# 1.0.0
First release

lib/main.dart
import 'package:flutter/material.dart';
import 'package:sha_alert/button.dart';
import 'package:sha_alert/sha_alert.dart';

void main() {
  runApp(const ShaAlertApp());
}

class ShaAlertApp extends StatelessWidget {
  const ShaAlertApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      title: "Sha Alert",
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  const Home({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Center(
          child: Button(
            onPressed: () {
              ShaAlert.showHozzoAlert(
                context,
                title: "Welcome Note",
                message: "Happy to see you on board! Please read the terms and conditions carefully before proceeding further.",
                firstButtonTitle: "I Agree",
                firstButtonType: ButtonType.primary,
                secondButtonTitle: "Read Terms and Conditions",
              );
            },
            title: "Delete Account",
            primaryColor: Colors.red,
            width: 200,
            height: 50,
          ),
        ),
      ),
    );
  }
}


lib/button.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:sha_alert/color.dart';
import 'package:sha_alert/const.dart';

class Button extends StatelessWidget {
  final String? title;
  final IconData? icon;
  final VoidCallback? onPressed;
  final bool isDisabled;
  final TextStyle? textStyle;
  final ButtonType buttonType;
  final bool isLoading;
  final double width;
  final double? height;
  final Color? secondaryColor;
  final Color? primaryColor;
  final Color? errorColor;
  final Color? successColor;
  const Button({
    super.key,
    this.title,
    this.width = 100,
    this.onPressed,
    this.icon,
    this.isLoading = false,
    this.isDisabled = false,
    this.textStyle,
    this.buttonType = ButtonType.primary,
    this.height,
    this.secondaryColor,
    this.primaryColor,
    this.errorColor,
    this.successColor,
  });

  @override
  Widget build(BuildContext context) {
    return InkWell(
      mouseCursor: isDisabled
          ? SystemMouseCursors.forbidden
          : isLoading
              ? SystemMouseCursors.wait
              : SystemMouseCursors.click,
      onTap: () {
        if (onPressed != null && !isDisabled && !isLoading) {
          onPressed!();
        }
      },
      child: Container(
        height: height,
        padding: const EdgeInsets.all(10),
        width: width,
        decoration: BoxDecoration(
          color: isDisabled
              ? const Color.fromARGB(26, 121, 121, 121)
              : buttonType == ButtonType.primary
                  ? primaryColor
                  : buttonType == ButtonType.secondary
                      ? secondaryColor
                      : buttonType == ButtonType.positive
                          ? successColor
                          : buttonType == ButtonType.negative
                              ? errorColor
                              : buttonType == ButtonType.warning
                                  ? UiColors.hexToColor('#FFC107')
                                  : primaryColor,
          borderRadius: BorderRadius.circular(ThemeConst.defaultBorderRadiusSmall),
        ),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          mainAxisSize: MainAxisSize.min,
          children: [
            const Spacer(),
            if (isLoading) const CupertinoActivityIndicator(color: Colors.white),
            if (!isLoading && title != null)
              Text(
                title!,
                style: textStyle ??
                    TextStyle(
                      color: isDisabled ? UiColors.hexToColor('#000000') : Colors.white,
                      fontSize: 15,
                      fontWeight: FontWeight.w500,
                    ),
              ),
            if (icon != null && !isLoading)
              Padding(
                padding: const EdgeInsets.symmetric(horizontal: 5),
                child: Icon(icon, color: isDisabled ? UiColors.hexToColor('#000000') : textStyle?.color ?? Colors.white, size: 16),
              ),
            if (!isLoading) const Spacer(),
          ],
        ),
      ),
    );
  }
}

//Types of buttons:
enum ButtonType {
  primary,
  secondary,
  positive,
  negative,
  warning,
}


linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "sha_alert");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "sha_alert");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"sha_alert", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/sha_alert.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:sha_alert/button.dart';

class ShaAlert {
  static showHozzoAlert(
    //* Required parameters
    //* This context should contain navigator for the alert to work properly and pop itself.
    BuildContext context, {
    required String title,
    required String message,
    required String firstButtonTitle,

    //* Second button title, if not provided, only one button will be shown
    String? secondButtonTitle,

    //* Function to be called when first button is pressed
    Function? onFirstButtonPressed,

    //* Function to be called when second button is pressed
    Function? onSecondButtonPressed,
    ButtonType firstButtonType = ButtonType.negative,
    ButtonType secondButtonType = ButtonType.secondary,

    //* Height of the alert dialog, default is 270 pixels.
    double height = 280,

    //* Use this for showing custom widgets or even loti or images, but don't forget to set height accordingly
    Widget? child,

    //* Backdrop color, default is black54 (black with 54% opacity)
    Color backdropColor = Colors.black54,

    //* Default transition duration is 120 milliseconds (0.12 seconds) but you can change it to whatever you want (120 is recommended)
    Duration transitionDuration = const Duration(milliseconds: 120),

    //* Function to be called when alert is dismissed
    Function? onDismissed,

    //* Custom haptic feedback function, if not provided, default haptic feedback will be used (Called after 100 milliseconds)
    Function? customHapticFeedback,
    Color? secondaryColor,
    Color? primaryColor,
    Color? errorColor,
    Color? successColor,
    bool overrideTheme = true,
  }) {
    // Setting default colors
    if (overrideTheme == false) {
      primaryColor ??= Theme.of(context).primaryColor;
      secondaryColor ??= Theme.of(context).colorScheme.secondary;
      errorColor ??= Theme.of(context).colorScheme.error;
      successColor ??= Theme.of(context).colorScheme.primary;
    } else {
      primaryColor ??= Colors.blue;
      secondaryColor ??= Colors.grey.shade100;
      errorColor ??= Colors.red;
      successColor ??= Colors.green;
    }

    //* Haptic feedback if available on device
    //* Only plays after 200 milliseconds for better user experience :).
    Future.delayed(const Duration(milliseconds: 100), () => customHapticFeedback != null ? customHapticFeedback() : HapticFeedback.lightImpact());
    showGeneralDialog(
      context: context,
      barrierLabel: "Sha Alert",
      barrierDismissible: true,
      barrierColor: backdropColor,
      transitionDuration: transitionDuration,
      useRootNavigator: true,
      pageBuilder: (_, __, ___) {
        return Material(
          type: MaterialType.transparency,
          child: Center(
            child: Container(
                height: message.length > 80 ? height + 50 : height,
                margin: const EdgeInsets.symmetric(horizontal: 20),
                decoration: BoxDecoration(color: Colors.white, borderRadius: BorderRadius.circular(30), boxShadow: [BoxShadow(color: Colors.black.withOpacity(0.1), blurRadius: 10, offset: const Offset(0, 10))]),
                child: SizedBox.expand(
                  child: Column(
                    children: [
                      const SizedBox(height: 30),
                      Text(title, style: Theme.of(context).textTheme.displayLarge?.copyWith(fontSize: 22, fontWeight: FontWeight.bold, color: Colors.black)),
                      const SizedBox(height: 15),
                      SizedBox(
                        width: 300,
                        child: Text(
                          message,
                          style: TextStyle(color: Colors.black.withOpacity(0.7), fontSize: 15),
                          textAlign: TextAlign.center,
                        ),
                      ),

                      // Custom child widget, if any
                      if (child != null) child,

                      // Action buttons
                      const SizedBox(height: 5),
                      Expanded(
                        child: Padding(
                          padding: EdgeInsets.symmetric(vertical: (secondButtonTitle != null && secondButtonTitle != "") ? 34 : 0, horizontal: 20),
                          child: Column(
                            crossAxisAlignment: CrossAxisAlignment.stretch,
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: [
                              // If only one button is provided, show this spacer
                              Visibility(visible: (secondButtonTitle == null || secondButtonTitle == ""), maintainState: false, child: const Spacer()),
                              // First button
                              Expanded(
                                child: Button(
                                  primaryColor: primaryColor,
                                  secondaryColor: secondaryColor,
                                  errorColor: errorColor,
                                  successColor: successColor,
                                  title: firstButtonTitle,
                                  textStyle: firstButtonType == ButtonType.secondary ? const TextStyle(color: Colors.black) : null,
                                  onPressed: () {
                                    HapticFeedback.lightImpact();
                                    if (onFirstButtonPressed != null) {
                                      onFirstButtonPressed();
                                    } else {
                                      Navigator.pop(context);
                                    }
                                  },
                                  width: double.infinity,
                                  buttonType: firstButtonType,
                                ),
                              ),

                              // Spacer if second button is not provided
                              Visibility(visible: (secondButtonTitle == null || secondButtonTitle == ""), maintainState: false, child: const Spacer()),

                              // If Both buttons are provided, show this spacer
                              Visibility(visible: (secondButtonTitle != null && secondButtonTitle != ""), maintainState: false, child: const SizedBox(height: 10)),

                              Visibility(
                                visible: (secondButtonTitle != null && secondButtonTitle != ""),
                                maintainState: false,
                                child: Expanded(
                                  child: Button(
                                    primaryColor: primaryColor,
                                    secondaryColor: secondaryColor,
                                    errorColor: errorColor,
                                    successColor: successColor,
                                    title: secondButtonTitle,
                                    onPressed: () {
                                      HapticFeedback.lightImpact();
                                      if (onSecondButtonPressed != null) {
                                        onSecondButtonPressed();
                                      } else {
                                        Navigator.pop(context);
                                      }
                                    },
                                    width: double.infinity,
                                    buttonType: secondButtonType,
                                    textStyle: const TextStyle(color: Colors.black),
                                  ),
                                ),
                              ),
                            ],
                          ),
                        ),
                      ),
                    ],
                  ),
                )),
          ),
        );
      },
      transitionBuilder: (_, anim, __, child) {
        // Animation should start at 0.5 and end at 1.0
        final tween = Tween<double>(begin: 0.85, end: 1.0);
        anim = tween.animate(anim);
        anim = CurvedAnimation(parent: anim, curve: Curves.fastOutSlowIn);

        return ScaleTransition(
          scale: anim,
          child: FadeTransition(
            opacity: anim,
            child: child,
          ),
        );
      },
    );
  }
}
